src/dotnet-roslyn-tools/DartTest/DartTest.cs

```diff
@@ -0,0 +1,237 @@
+                return -1;
+            }
+            else
+            {
+                logger.LogError("A SHA is required for the given product.");
+                return -1;
+            }
+        }
+
+        var directory = await PushPRToInternal(product, prNumber, azureBranchName, logger, sha, cancellationToken).ConfigureAwait(false);
+        var repositoryParams = new Dictionary<string, RepositoryResourceParameters>
+            {
+                {
+                    "self", new RepositoryResourceParameters
+                    {
+                        RefName = $"refs/heads/{azureBranchName}",
+                        Version = sha
+                    }
+                }
+            };
+
+        var runPipelineParameters = new RunPipelineParameters
+        {
+            Resources = new RunResourcesParameters
+            {
+
+            },
+            TemplateParameters = new Dictionary<string, string> { { "prNumber", prNumber.ToString() }, { "sha", sha } }
+        };
+
+        await remoteConnections.DevDivConnection.TryRunPipelineAsync(product.DartLabPipelineName, repositoryParams, runPipelineParameters, logger).ConfigureAwait(false);
+        CleanupDirectory(directory, logger);
+        return 0;
+    }
+
+    private static async Task<string?> GetLatestShaFromPullRequest(IProduct product, HttpClient gitHubClient, int prNumber, ILogger logger, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var requestUri = $"/repos/dotnet/{product.Name.ToLower()}/pulls/{prNumber}";
+            var response = await gitHubClient.GetAsync(requestUri, cancellationToken).ConfigureAwait(false);
+
+            if (response.IsSuccessStatusCode)
+            {
+                var prData = await response.Content.ReadFromJsonAsync<JsonObject>(cancellationToken: cancellationToken).ConfigureAwait(false);
+                return prData?["head"]?["sha"]?.ToString();
+            }
+            else
+            {
+                logger.LogError($"Failed to retrieve PR data from GitHub. Status code: {response.StatusCode}");
+            }
+        }
+        catch (Exception e)
+        {
+            logger.LogError(e, "Exception while retrieving the latest SHA from GitHub.");
+        }
+
+        return null;
+    }
+
+
+    private static async Task<string?> PushPRToInternal(IProduct product, int prNumber, string azureBranchName, ILogger logger, string sha, CancellationToken cancellationToken)
+    {
+        string? targetDirectory = null;
+
+        try
+        {
+            targetDirectory = Path.Combine(Path.GetTempPath(), $"pr-{prNumber}");
+            var counter = 1;
+            while (Directory.Exists(targetDirectory))
+            {
+                targetDirectory = Path.Combine(Path.GetTempPath(), $"pr-{prNumber}-{counter}");
+                counter++;
+            }
+
+            Directory.CreateDirectory(targetDirectory);
```

Comment:
Maybe pull this bit out so that the temp directory is created and cleaned up by the same method.